pkg/controller/direct/logging/link_controller.go

```diff
@@ -0,0 +1,238 @@
+
+	gcpClient, err := gcp.NewConfigRESTClient(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("building Logging Config client: %w", err)
+	}
+	return gcpClient, err
+}
+
+func (m *modelLoggingLink) AdapterForObject(ctx context.Context, reader client.Reader, u *unstructured.Unstructured) (directbase.Adapter, error) {
+	obj := &krm.LoggingLink{}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, &obj); err != nil {
+		return nil, fmt.Errorf("error converting to %T: %w", obj, err)
+	}
+
+	id, err := krm.NewLoggingLinkRef(ctx, reader, obj)
+	if err != nil {
+		return nil, err
+	}
+
+	gcpClient, err := m.client(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return &LoggingLinkAdapter{
+		id:        id,
+		gcpClient: gcpClient,
+		desired:   obj,
+	}, nil
+}
+
+/*
+func () ResolveExternalRef(externalRef string) {
+
+}
+*/
+
+func (m *modelLoggingLink) AdapterForURL(ctx context.Context, url string) (directbase.Adapter, error) {
+	// TODO: Support URLs
+	return nil, nil
+}
+
+type LoggingLinkAdapter struct {
+	id        *krm.LoggingLinkRef
+	gcpClient *gcp.ConfigClient
+	desired   *krm.LoggingLink
+	actual    *loggingpb.Link
+}
+
+var _ directbase.Adapter = &LoggingLinkAdapter{}
+
+func (a *LoggingLinkAdapter) Find(ctx context.Context) (bool, error) {
+	log := klog.FromContext(ctx)
+	log.V(2).Info("getting LoggingLink", "name", a.id.External)
+
+	req := &loggingpb.GetLinkRequest{Name: a.id.External}
+	linkpb, err := a.gcpClient.GetLink(ctx, req)
+	if err != nil {
+		if direct.IsNotFound(err) {
+			return false, nil
+		}
+		return false, fmt.Errorf("getting LoggingLink %q: %w", a.id.External, err)
+	}
+
+	a.actual = linkpb
+	return true, nil
+}
+
+func (a *LoggingLinkAdapter) Create(ctx context.Context, createOp *directbase.CreateOperation) error {
+	u := createOp.GetUnstructured()
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("creating object", "u", u)
+
+	mapCtx := &direct.MapContext{}
+
+	desired := a.desired.DeepCopy()
+	resource := LoggingLinkSpec_ToProto(mapCtx, &desired.Spec)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+
+	// TODO(user): Complete the gcp "CREATE" or "INSERT" request with required fields.
+	// TODO(): 400 error needs to come through on the error
+	// TODO(): there is an implicit dependency on the bucket being active, do we need that here?
+	parent, err := a.id.Parent()
+	if err != nil {
+		return err
+	}
+
+	resourceID := direct.ValueOf(desired.Spec.ResourceID)
+	if resourceID == "" {
+		log.V(2).Info("ResourceID is not set, will use metadata.name")
+		resourceID = desired.Name
+	} else {
+		log.V(2).Info("ResourceID is set, will use")
+		resourceID = *desired.Spec.ResourceID
+	}
+
+	req := &loggingpb.CreateLinkRequest{
+		Parent: parent.String(),
+		Link:   resource,
+		LinkId: resourceID,
+	}
+	op, err := a.gcpClient.CreateLink(ctx, req)
+	if err != nil {
+		return fmt.Errorf("creating Link %s: %w\n", a.id.External, err)
```

Comment:
`a.id` should be good as it'll automatically call its String() function when resolving the error message. All the usage of `a.id.External` in string value resolution should be replaced with `a.id`.